src/linter/ui5Types/SourceFileLinter.ts

```diff
@@ -124,81 +124,118 @@ export default class SourceFileLinter {
+		} else if (
+			ts.isPropertyDeclaration(node) && node.name.getText() === "metadata" &&
+			ts.isClassDeclaration(node.parent) && this.isUi5ClassDeclaration(node.parent, "sap/ui/base/ManagedObject")
+		) {
 			const visitMetadataNodes = (childNode: ts.Node) => {
 				if (ts.isPropertyAssignment(childNode)) { // Skip nodes out of interest
 					this.analyzeMetadataProperty(childNode.name.getText(), childNode);
 				}
 
 				ts.forEachChild(childNode, visitMetadataNodes);
 			};
-
-			if (this.isUi5controlMetadataNode(node)) {
-				ts.forEachChild(node, visitMetadataNodes);
-			}
+			ts.forEachChild(node, visitMetadataNodes);
+		} else if (ts.isClassDeclaration(node) && this.isUi5ClassDeclaration(node, "sap/ui/core/Control")) {
+			this.analyzeControlRendererDeclaration(node);
 		}
 
 		// Traverse the whole AST from top to bottom
 		ts.forEachChild(node, this.#boundVisitNode);
 	}
 
-	isUi5controlMetadataNode(node: ts.PropertyAssignment | ts.PropertyDeclaration): boolean {
-		// Go up the hierarchy chain to find whether the class extends from "sap/ui/base/ManagedObject"
-		const isObjectMetadataAncestor = (node: ts.ClassDeclaration): boolean => {
+	isUi5ClassDeclaration(node: ts.ClassDeclaration, baseClassModule: string | string[]): boolean {
+		const baseClassModules = Array.isArray(baseClassModule) ? baseClassModule : [baseClassModule];
+		const baseClasses = baseClassModules.map((baseClassModule) => {
+			return {module: baseClassModule, name: baseClassModule.split("/").pop()};
+		});
+
+		// Go up the hierarchy chain to find whether the class extends from the provided base class
+		const isClassUi5Subclass = (node: ts.ClassDeclaration): boolean => {
 			return node?.heritageClauses?.flatMap((parentClasses: ts.HeritageClause) => {
 				return parentClasses.types.flatMap((parentClass) => {
 					const parentClassType = this.#checker.getTypeAtLocation(parentClass);
 
 					return parentClassType.symbol?.declarations?.flatMap((declaration) => {
-						if (ts.isClassDeclaration(declaration)) {
-							if (declaration.name?.getText() === "ManagedObject" &&
+						if (!ts.isClassDeclaration(declaration)) {
+							return false;
+						}
+						for (const baseClass of baseClasses) {
+							if (declaration.name?.getText() === baseClass.name &&
 								(
-									// Declaration via type definitions
+								// Declaration via type definitions
 									(
 										declaration.parent.parent &&
 										ts.isModuleDeclaration(declaration.parent.parent) &&
-										declaration.parent.parent.name?.text === "sap/ui/base/ManagedObject"
+										declaration.parent.parent.name?.text === baseClass.module
 									) ||
 									// Declaration via real class (within sap.ui.core project)
 									(
 										ts.isSourceFile(declaration.parent) &&
-										declaration.parent.fileName === "/resources/sap/ui/base/ManagedObject.js"
+										declaration.parent.fileName === `/resources/${baseClass.module}.js`
```

Comment:
Does currently not happen for sap.ui.core code, but what if the code would be in TypeScript in future? (*.ts)
Could happen if we ever introduce a check against a sap.fe base class?

Reply:
Good catch. I will also add a `.ts` extension check.